cocotb_test/simulator.py

```diff
@@ -306,15 +340,78 @@ def exit_gracefully(self, signum, frame):
         signal.signal(signal.SIGTERM, self.old_sigterm_h)
         assert False, "Exiting pid: {} with signum: {}".format(str(pid), str(signum))
 
+    def format_input(self):
+        # format sources input
+        if self.has_namedlib:
+            if self.vhdl_sources:
+                if isinstance(self.vhdl_sources, dict):
+                    self.use_namedlib = True
+                else:
+                    self.vhdl_sources = {f"{self.toplevel_name}": self.vhdl_sources}
+
+            if self.verilog_sources:
+                if isinstance(self.verilog_sources, dict):
+                    self.use_namedlib = True
+                else:
+                    self.verilog_sources = {f"{self.toplevel_name}": self.verilog_sources}
+        else:
+            if self.vhdl_sources is not None:
+                assert isinstance(self.vhdl_sources, list), "Parameter `vhdl_sources` must be a list."
+            if self.verilog_sources is not None:
+                assert isinstance(self.verilog_sources, list), "Parameter `verilog_sources` must be a list."
+
+        # format toplevel input
+        if self.has_multitop:
+            if isinstance(self.toplevel, list):
+                self.use_multitop = True
+            else:
+                self.toplevel = [self.toplevel]
+        else:
+            assert isinstance(self.toplevel, str), "Parameter `toplevel` must be a string."
+
+        # check that every toplevel specifies library when using namedlib
+        if self.use_namedlib:
+            if self.has_multitop:
+                toplevel = self.toplevel
+            else:
+                toplevel = [self.toplevel]
+            for v in toplevel:
+                assert "." in v, "When using named libraries, toplevels must be specified as '<library>.<module>'."
+
+        # format toplevel inuput when not using namedlib
+        if self.has_namedlib and not self.use_namedlib:
+            default_lib = self.toplevel[0]
+            if self.has_multitop:
+                for j, v in enumerate(self.toplevel):
+                    if not "." in v:
+                        self.toplevel[j] = default_lib + "." + v
+            else:
+                if not "." in self.toplevel:
+                    self.toplevel = default_lib + "." + self.toplevel
+
+    @property
+    def toplevel_first(self) -> str:
+        """First toplevel entry."""
+        if isinstance(self.toplevel, list):
+            return self.toplevel[0]
+        return self.toplevel
+
+    @property
+    def toplevel_name(self):
```

Comment:
I would want to suggest `toplevel_first_name` for consistency, but then it sounds stupid...
Change to `first_toplevel(_name)` maybe?

Reply:
would need a property `toplevel_last_name` then, too ;).

Joking aside, I think consistency is more important than aesthetics.

We could also instead of `toplevel_first` and `toplevel_first_name` create properties which return indexable module names which can be accessed like `self.toplevel`:

```python
@property
def toplevel_module(self):
    """Returns toplevel module name[s]."""
    if isinstance(self.toplevel, list):
        return [s.rsplit(".", 1)[-1] for s in self.toplevel]
    else:
        return self.toplevel.rsplit(".", 1)[-1]
```
This way, `self.toplevel_module` always has the same data type and index as `self.toplevel`. Analogue to that we could create a `self.toplevel_library` property.